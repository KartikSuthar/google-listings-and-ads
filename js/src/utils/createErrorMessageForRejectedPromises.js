/**
 * Internal dependencies
 */
import createNoticeForMultipleErrors from '.~/utils/createMessageForMultipleErrors';

/**
 * A function to create an error message for rejected promises.
 *
 * @param  {Array<Promise>} promises Promises to be called.
 * @param  {Array<string>} promiseNames Name of each provided promise in the first argument. It will be used to create the error message and needs to follow the same order that in the first argument.
 * @return {Promise<string|null>} - The error message for rejected promises otherwise null if there are not rejected promises.
 */
export default async function createErrorMessageForRejectedPromises(
	promises,
	promiseNames
) {
	const results = await Promise.allSettled( promises );
	const rejectedMessages = results.reduce( ( acc, item, index ) => {
		if ( item.status === 'rejected' && promiseNames[ index ] ) {
			acc.push( promiseNames[ index ] );
		}
		return acc;
	}, [] );

	const isPartiallySuccessful = rejectedMessages.length < promises.length;

	return createNoticeForMultipleErrors(
		rejectedMessages,
		isPartiallySuccessful
	);
}
