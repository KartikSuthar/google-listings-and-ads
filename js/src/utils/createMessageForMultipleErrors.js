/**
 * External dependencies
 */
import { __, sprintf, _n } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import concatenateListOfWords from '.~/utils/concatenateListOfWords';

/**
 * Creates a message for a list of error messages. For example for a list of error messages: createMessageForMultipleErrors(['Target Audience', 'Shipping Rates']) will create the following message:
 * 'There are errors in the following actions: Target Audience and Shipping Rates. Please try again later.'
 *
 * @param {Array<string>} errorMessages Error messages to be displayed in the notice.
 * @param {boolean} [isPartiallySuccessful='true'] Whether some actions were successful.
 * @return {string|null} - Return the error message for a list of error otherwise null if there are not error messages
 */
export default function createMessageForMultipleErrors(
	errorMessages,
	isPartiallySuccessful = true
) {
	if ( errorMessages.length ) {
		const listErrors = concatenateListOfWords( errorMessages );

		const content = sprintf(
			// translators: 1: list of errors 2: optional text if some promises have been invoked successfully.
			_n(
				'There is an error in the following action: %1$s. %2$s Please try again later.',
				'There are errors in the following actions: %1$s. %2$s Please try again later.',
				errorMessages.length,
				'google-listings-and-ads'
			),
			listErrors,
			isPartiallySuccessful ? __( 'Other changes have been saved.' ) : ''
		);

		return content;
	}

	return null;
}
