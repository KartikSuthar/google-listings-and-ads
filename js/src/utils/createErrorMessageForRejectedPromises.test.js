/**
 * Internal dependencies
 */
import createErrorMessageForRejectedPromises from './createErrorMessageForRejectedPromises';
import createMessageForMultipleErrors from '.~/utils/createMessageForMultipleErrors';

jest.mock( '.~/utils/createMessageForMultipleErrors', () => {
	const originalModule = jest.requireActual(
		'.~/utils/createMessageForMultipleErrors'
	);
	return jest.fn( originalModule.default );
} );

describe( 'createErrorMessageForRejectedPromises', () => {
	const successPromise = () => new Promise( ( resolve ) => resolve( 'OK' ) );
	const rejectedPromise = () =>
		new Promise( ( r, reject ) => reject( 'Failed' ) );

	afterEach( () => {
		jest.clearAllMocks();
	} );

	it( 'Called with no rejected promises, calls `createMessageForMultipleErrors` with an empty array and `true`, and returns `null`', async () => {
		const promises = [
			successPromise(),
			successPromise(),
			successPromise(),
		];
		const promisesName = [ 'Promise A', 'Promise B', 'Promise C' ];

		const message = await createErrorMessageForRejectedPromises(
			promises,
			promisesName
		);

		expect( message ).toBeNull();
		expect( createMessageForMultipleErrors ).toBeCalledTimes( 1 );
		expect( createMessageForMultipleErrors ).toBeCalledWith( [], true );
	} );

	it( 'Called with one rejected promise, calls `createMessageForMultipleErrors` with its name, and `true`, and returns a message indicates it contains one error and partially successful results', async () => {
		const promises = [
			successPromise(),
			rejectedPromise(),
			successPromise(),
		];
		const promisesName = [ 'Promise A', 'Promise B', 'Promise C' ];

		const message = await createErrorMessageForRejectedPromises(
			promises,
			promisesName
		);

		expect( message ).toMatchSnapshot();
		expect( createMessageForMultipleErrors ).toBeCalledTimes( 1 );
		expect( createMessageForMultipleErrors ).toBeCalledWith(
			[ 'Promise B' ],
			true
		);
	} );

	it( 'Called with multiple rejected promises, calls `createMessageForMultipleErrors` with their names, and `true`, and returns a message indicates it contains multiple errors and partially successful results', async () => {
		const promises = [
			successPromise(),
			rejectedPromise(),
			rejectedPromise(),
			rejectedPromise(),
		];
		const promisesName = [
			'Promise A',
			'Promise B',
			'Promise C',
			'Promise D',
		];

		const message = await createErrorMessageForRejectedPromises(
			promises,
			promisesName
		);

		expect( message ).toMatchSnapshot();
		expect( createMessageForMultipleErrors ).toBeCalledTimes( 1 );
		expect( createMessageForMultipleErrors ).toBeCalledWith(
			[ 'Promise B', 'Promise C', 'Promise D' ],
			true
		);
	} );

	it( 'Called with all rejected promises, calls `createMessageForMultipleErrors` with their names, and `false`, and returns a message indicates it contains multiple errors without successful results', async () => {
		const promises = [
			rejectedPromise(),
			rejectedPromise(),
			rejectedPromise(),
		];
		const promisesName = [ 'Promise A', 'Promise B', 'Promise C' ];

		const message = await createErrorMessageForRejectedPromises(
			promises,
			promisesName
		);

		expect( message ).toMatchSnapshot();
		expect( createMessageForMultipleErrors ).toBeCalledTimes( 1 );
		expect( createMessageForMultipleErrors ).toBeCalledWith(
			[ 'Promise A', 'Promise B', 'Promise C' ],
			false
		);
	} );
} );
