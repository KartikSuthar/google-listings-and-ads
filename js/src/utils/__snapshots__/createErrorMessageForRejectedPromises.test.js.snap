// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createErrorMessageForRejectedPromises Called with all rejected promises, calls \`createMessageForMultipleErrors\` with their names, and \`false\`, and returns a message indicates it contains multiple errors without successful results 1`] = `"There are errors in the following actions: Promise A, Promise B and Promise C. Please try again later."`;

exports[`createErrorMessageForRejectedPromises Called with multiple rejected promises, calls \`createMessageForMultipleErrors\` with their names, and \`true\`, and returns a message indicates it contains multiple errors and partially successful results 1`] = `"There are errors in the following actions: Promise B, Promise C and Promise D. Other changes have been saved. Please try again later."`;

exports[`createErrorMessageForRejectedPromises Called with one rejected promise, calls \`createMessageForMultipleErrors\` with its name, and \`true\`, and returns a message indicates it contains one error and partially successful results 1`] = `"There is an error in the following action: Promise B. Other changes have been saved. Please try again later."`;
